File: clojure/robot-simulator/src/robot_simulator.clj
(ns robot-simulator)

(def ^:private bearings
  {:north {:advance-coord :y :advance-op inc :left :west  :right :east}
   :east  {:advance-coord :x :advance-op inc :left :north :right :south}
   :south {:advance-coord :y :advance-op dec :left :east  :right :west}
   :west  {:advance-coord :x :advance-op dec :left :south :right :north}})

(defn robot [coordinates bearing]
  {:coordinates coordinates :bearing bearing})

(def turn-right (comp :right bearings))
(def turn-left (comp :left bearings))

(defn- advance [bearing coordinates]
  (let [{:keys [advance-coord advance-op]} (bearings bearing)]
    (update coordinates advance-coord advance-op)))

(defn- apply-instruction [robot instruction]
  (case instruction
    \R (update robot :bearing turn-right)
    \L (update robot :bearing turn-left)
    \A (update robot :coordinates (partial advance (:bearing robot)))
    (assert false (str "Invalid instruction: " instruction))))

(defn simulate [instructions robot]
  (reduce apply-instruction robot instructions))

File: python/robot-simulator/_iterations/1/robot_simulator.py
from functools import reduce

EAST = complex(1, 0)
NORTH = complex(0, 1)
WEST = complex(-1, 0)
SOUTH = complex(0, -1)

class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.position = complex(x_pos, y_pos)
        self.direction = direction

    @property
    def coordinates(self):
        return self.position.real, self.position.imag

    def move(self, instructions):
        self.position, self.direction = reduce(
            lambda state, instruction: MOVES[instruction](*state),
            instructions,
            (self.position, self.direction),
        )

MOVES = {
    'L': lambda position, direction: (position, direction * 1j),
    'R': lambda position, direction: (position, direction * -1j),
    'A': lambda position, direction: (position + direction, direction),
}
